File: sicp_1_21.swift
#!/usr/bin/env swift
// sicp exercise 1.21

func square(x: Int) -> Int {
  return x * x
}

func smallestDivisor(x: Int) -> Int {
  return findDivisor(x,2)
}

func findDivisor(n: Int, t: Int) -> Int {
  return square(t) > n ? n : divides(t,n) ? t : findDivisor(n,(t+1))
}

func divides(a: Int, b: Int) -> Bool {
  return b % a == 0
}

func prime(n: Int) -> Bool {
  return smallestDivisor(n) == n
}

println(smallestDivisor(199))
println(smallestDivisor(1999))
println(smallestDivisor(19999))

File: sicp_1_21.js
#!/usr/bin/env node

function square(x) { return x * x }
function smallestDivisor(x) { return findDivisor(x,2) }
function divides(x,y) { return y % x === 0 }

function findDivisor(n,testDivisor) {
  if (square(testDivisor) > n) {
    return n;
  } else if (divides(testDivisor,n)) {
    return testDivisor;
  } else {
    return findDivisor(n,(testDivisor+1));
  }
}

function prime(x) { return smallestDivisor(x) === x };

console.log(smallestDivisor(199));
console.log(smallestDivisor(1999));
console.log(smallestDivisor(19999));

for (var i = 2; i <= 100; i++) {
  if (prime(i)) { console.log(i) };
}